File: Python/Programmers/Leve1/약수의 개수와 덧셈.py
# def solution(left, right):
#     answer = 0
#     for num in range(left, right+1):
#         cnt = 0
#         for i in range(1, num+1): # 1 ~ num까지 약수의 개수를 구함
#             if num % i == 0: # 약수면
#                 cnt += 1 # count
#         if cnt % 2 == 0: # 짝수면 더하고
#             answer += num
#         else: # 홀수면 뺄셈
#             answer -= num
    
#     return answer

def solution(left, right):
    answer = 0
    for num in range(left, right+1):
        if int(num**0.5) == num**0.5: # num이 제곱근이 정수이면 약수의 개수는 홀수
            answer -= num
        else:
            answer += num

    return answer

File: JavaScript/Programmers/Level 1/약수의 개수와 덧셈.js
// 약수의 개수는 1은 1개, 2는 2개, 3은 2개, 4는 3개, 5는 2개, 6은 4개, 7은 2개, 8은 4개, 9는 3개, ...
// 즉, n의 제곱근이 정수이면 약수의 개수는 홀수, 나머지는 짝수가 됨

// function solution(left, right) {
//     let answer = 0;
//     for (let num = left; num <= right; num++) {
//         // Number.isInteger 메소드를 이용하여 제곱근이 정수인지 확인
//         answer += (Number.isInteger(num ** 0.5) ? -num : num); // 제곱근이 정수이면 뺄셈, 정수가 아니면 덧셈
//     }
    
//     return answer;
// }

// answer 배열에 left부터 right까지 모든 숫자를 넣고 reduce 메소드를 이용하여 return

// function solution(left, right) {
//     let answer = [];
//     for (let num = left; num <= right; num++) {
//         answer.push(num); // left ~ right까지 모든 숫자 push
//     }
    
//     return answer.reduce((accumulator, current)=> { // reduce 메소드를 이용하여 값을 구함
//         return accumulator + (Number.isInteger(current ** 0.5) ? -current : current); // 제곱근이 정수이면 뺄셈, 정수가 아니면 
//     }, 0);
// }

// Array를 이용해서 배열을 만들고, fill 메소드를 이용해서 초기값 설정 후 reduce 메소드 사용
function solution(left, right) {
    return Array(right - left + 1)
        .fill(0)
        .reduce((answer, _, idx) => answer + (Number.isInteger((idx + left) ** 0.5) ? -(idx + left) : (idx + left)), 0);
}
